Multivariate functions

benjione opened the pull request:
Multivariate functions for Polynomial spaces (and maybe others).

Work in progress

benjione commented:
current implementation, compared to the 2D case which already existed and still uses 100% the old code:
```
julia> f2 = Fun(Chebyshev()^2, rand(100));

julia> f3 = Fun(Chebyshev()^3, rand(100));

julia> @btime f2(0.2,0.4);
  261.889 μs (3692 allocations: 178.70 KiB)

julia> @btime f3(0.2,0.4,0.5);
  400.441 μs (6054 allocations: 244.28 KiB)
```
Most time is needed to change the coefficients format from #86 to tensor format. Maybe it would be better to change the vectorization after all.

benjione commented:
The PR is not finished yet, I haven't added tests yet and will probably have to optimize some parts.

jishnub commented:
Ok, I've cancelled the tests for now to avoid clogging CI resources

benjione commented:
reduced amount of allocations and increased performance: 

```
julia> f2 = Fun(Chebyshev()^2, rand(100));

julia> f3 = Fun(Chebyshev()^3, rand(100));

julia> f4 = Fun(Chebyshev()^4, rand(100));

julia> @btime f2(0.2, 0.4);
  261.099 μs (3725 allocations: 177.47 KiB)

julia> @btime f3(0.2, 0.4, 0.5);
  379.646 μs (4781 allocations: 206.83 KiB)

julia> @btime f3(0.2, 0.4, 0.5);
  379.589 μs (4781 allocations: 206.83 KiB)

julia> @btime f4(0.2, 0.4, 0.5, 0.6);
  474.031 μs (5180 allocations: 247.55 KiB)

```

benjione commented:
The basic functionality of multivariate functions of the same type is now working.

```
julia> f3 = Fun(Chebyshev()^3, rand(100));

julia> f3(0.2, 0.4, 0.5); # working

julia> f3 = Fun(Chebyshev() \otimes Legendre() \otimes Chebyshev(), rand(100));

julia> f3(0.2, 0.4, 0.5); # not working
```
The mixed type of (polynomial) spaces should be easy to extend, but is beyond this PR in my opinion. The same thing for arithmetics, which also do not work fully in the 2D case #187 (I explicitly did not tuch the 2D case yet).

I think the PR is ready for a review or merge.

benjione commented:
From my side the PR is ready to be merged @jishnub

jishnub commented on test/MultivariateTest.jl:
```diff
@@ -0,0 +1,81 @@
+        end
+    end
+
+end
```
Could you add a newline at the end of the file?

jishnub commented:
How much does including `Combinatorics` add to the package import time?

jishnub commented on src/Multivariate/ProductFun.jl:
```diff
@@ -219,6 +238,33 @@ evaluate(f::ProductFun{S,V,SS,T},x::Number,y::Number) where {S<:UnivariateSpace,
+        tmp = f.coefficients[coef_counter]
+        if tmp != 0
+            tmp_res = 1
+            @inbounds @simd for k=1:d
```
Does `@simd` help here, since you're indexing into another iterator? Or is the `simd` applied to the inner indexing? If the performance gain isn't significant, perhaps it'll be better to leave out `@inbounds`

benjione replied:
I checked it, there exists a small performance difference, but it is small. Most time is spent in the iterator, I will delete it for now.

jishnub commented on src/Multivariate/ProductFun.jl:
```diff
@@ -5,6 +5,16 @@
+## however, refactoring this is a lot of effort...
+struct TensorIteratorFun{S<:UnivariateSpace, d, SS<:TensorSpace{NTuple{d, S}}, T<:Number} <: MultivariateFun{T, d}
+    space::SS
+    coefficients::AbstractVector{T} 
```
Do we need `AbstractVector` as the type? Or should this be concretely typed instead?

benjione replied:
changed

benjione commented on src/Multivariate/ProductFun.jl:
```diff
@@ -35,6 +45,15 @@ function ProductFun(cfs::AbstractMatrix{T},sp::AbstractProductSpace{Tuple{S,V},D
 
+## TODO: This Product Fun actually does not return a productfun, dirty but was easiest to implement. Probably an abstract type of ProductFuns
+# is needed in the future.
+function ProductFun(iter::TrivialTensorizer{d},cfs::Vector{T},blk::Block, sp::AbstractProductSpace{NTuple{d, S},DD}) where {S<:UnivariateSpace,T<:Number,d,DD}
```
Currently multivariate functions are designed in a way to first build a tensor out of the coefficients and then first solve the first dimension followed by the second. `ProductFun` is designed for that, storing a vector of functions.
I first implemented it that way, but this needs too much memory for high dimensional cases and threw errors for me for dimensions larger about 80 and 10000 coefficients (~1 million coefficients are possible in the 1D case for me). Therefore, I never build this tensor and don't use the vector of functions `ProductFun` is using. I think the approach is not compatible with the current `ProductFun`.

I think the best way would actually be to change the order of how the coefficients are vectorized, see #86 , because this change in coefficients is computationally costly (functions `fromtensor` and `totensor` build the "real" tensor out of the current way the coefficients are stored. I introduced `totensoriterator` for evaluating d>2 dimensional functions to avoid building the full tensor).

However, this change could be breaking with current code using ApproxFun

benjione commented:
How do I measure the import time exactly? I got with compilation the import time of
```
julia> @time import Combinatorics
  0.001712 seconds (1.08 k allocations: 72.438 KiB, 84.56% compilation time)
```
and precompiled in the package
```
@time using ApproxFunBase
  0.000143 seconds (127 allocations: 11.344 KiB)
```

jishnub commented:
You can try `@time_imports using ApproxFunBase` on Julia v1.8+

jishnub commented on src/Multivariate/TensorSpace.jl:
```diff
@@ -31,6 +31,74 @@ Base.eltype(::Tensorizer{NTuple{d,T}}) where {d,T} = NTuple{d,Int}
+end
+
+
+done(a::TrivialTensorizer{d}, ((block, (j, iterator, iter_state)), sub, shift, numblock, (i,tot))) where {d, AA} = i ≥ tot
```
The type parameter `AA` seems unused, tests probably fail because of this.

benjione commented:
> You can try `@time_imports using ApproxFunBase` on Julia v1.8+

I get
`5.8 ms  Combinatorics`

jishnub commented:
Could you post the full report? In particular, I want to see if including `Combinatorics` increases invalidation and impacts other packages. In itself, 5ms is hardly anything, so this shouldn't be an issue.

benjione commented:
```
julia> @time_imports using ApproxFunBase
    162.9 ms  FillArrays
    936.6 ms  ArrayLayouts
    181.6 ms  BlockArrays
    241.1 ms  BandedMatrices
    162.2 ms  MatrixFactorizations
     61.7 ms  BlockBandedMatrices
      2.9 ms  StaticArraysCore
    860.8 ms  StaticArrays
     34.0 ms  IntervalSets
      2.3 ms  CompositeTypes
    143.8 ms  DomainSets
     11.0 ms  AbstractFFTs
      0.8 ms  FFTW_jll
    305.1 ms  FFTW 6.88% compilation time (100% recompilation)
     45.9 ms  IterTools
    243.7 ms  Polynomials
     53.2 ms  DSP
      4.5 ms  Calculus
    139.6 ms  DualNumbers
     19.2 ms  Nullables
   1388.7 ms  LowRankApprox
     38.9 ms  Infinities
    163.0 ms  MacroTools
    133.3 ms  LazyArrays
    151.7 ms  InfiniteArrays
    214.7 ms  LazyBandedMatrices
     14.4 ms  SemiseparableMatrices
    952.0 ms  InfiniteLinearAlgebra
      6.7 ms  Combinatorics
    370.4 ms  ApproxFunBase

```

jishnub commented:
Ok, seems reasonable as it's fairly low in the chain. We can include this then

jishnub commented:
Tests seem to fail because of unbound type parameters in
```julia
ApproxFunBase.TensorIteratorFun(space::SS, coefficients::Vector{T}, iterator::ApproxFunBase.Tensorizer{Tuple{Vararg{FillArrays.Ones{Int64, 1, Tuple{InfiniteArrays.OneToInf{Int64}}}, d}}}, orders::BlockArrays.Block) where {S<:(Space{D} where D<:(Domain{<:Number})), d, SS<:(TensorSpace{Tuple{Vararg{S, d}}}), T<:Number} in ApproxFunBase at /Users/runner/work/ApproxFunBase.jl/ApproxFunBase.jl/src/Multivariate/ProductFun.jl:12, totensor(SS::TensorSpace{Tuple{Vararg{S, d}}}, M::AbstractVector) where {d, S<:(Space{D} where D<:(Domain{<:Number}))} in ApproxFunBase at /Users/runner/work/ApproxFunBase.jl/ApproxFunBase.jl/src/Multivariate/TensorSpace.jl:493

ProductFun(iter::ApproxFunBase.Tensorizer{Tuple{Vararg{FillArrays.Ones{Int64, 1, Tuple{InfiniteArrays.OneToInf{Int64}}}, d}}}, cfs::Vector{T}, blk::BlockArrays.Block, sp::ApproxFunBase.AbstractProductSpace{Tuple{Vararg{S, d}}, DD}) where {S<:(Space{D} where D<:(Domain{<:Number})), T<:Number, d, DD} in ApproxFunBase at /Users/runner/work/ApproxFunBase.jl/ApproxFunBase.jl/src/Multivariate/ProductFun.jl:50
```
Could you fix these?

benjione commented:
Thanks! I will fix these once I am back home next week.

jishnub commented:
Could you rebase on master? This might help with tests

benjione commented:
now it should work, at least when testing on my personal computer :smile:

jishnub commented:
Testing against `ApproxFunOrthogonalPolynomials` explicitly should not be necessary. Maybe I'll look into the issue here.

jishnub commented:
I see the issue here. This package is somewhat unconventional in the way the tests are set up. In this case, you'll have to open a PR with `ApproxFunOrthogonalPolynomials` and add the multivariate tests there, after a version has been tagged here without tests. Only some spaces such as `PointSpace` are available in this package, so see if you may add tests using this.

benjione commented:
Ok, I think I understand. The Multivariate methods I added only work for Univariate Tensor Spaces of the same type.

If I understand it correctly, I will delete the tests from here and move them to Orthogonal Polynomials. 

The problem is probably that if there is a braking change in this package, the test in the orthogonal polynomial package will break, not here. Is it possible to also include those tests in the github ci?

jishnub commented on src/Multivariate/ProductFun.jl:
```diff
@@ -74,6 +84,15 @@ function ProductFun(cfs::AbstractMatrix{T},sp::AbstractProductSpace{Tuple{S,V},D
 
+## TODO: This Product Fun actually does not return a productfun, dirty but was easiest to implement. Probably an abstract type of ProductFuns
+# is needed in the future.
+function ProductFun(iter::TrivialTensorizer{d},cfs::Vector{T},blk::Block, sp::AbstractProductSpace{<:NTuple{d, <:UnivariateSpace}}) where {T<:Number,d}
```
This should probably add a method to `MultivariateFun` instead of `ProductFun`

benjione replied:
I am not sure why, it just overwrites ProductFun for the TrivialTensorizer case.

jishnub replied:
A constructor for a struct should return a value of the same type. While it's possible to violate this, it's usually not a good idea. Do you specifically require `ProductFun` to me modified this way, to hit this method in the dispatch sequence somewhere?

benjione replied:
I will check if I can implement this in a better way. 
The problem is also that the ProductFun is only for bivariate functions and the implementation does not scale well to higher dimensions. Maybe one should completely separate the 2D case from higher dimensional cases or also rewrite the 2D case.
```
struct ProductFun{S<:UnivariateSpace,V<:UnivariateSpace,SS<:AbstractProductSpace,T} <: BivariateFun{T}
    coefficients::Vector{VFun{S,T}}     # coefficients are in x
    space::SS
end
```

Generally, the current order in which coefficients are stored is not ideal for the evaluation. In my opinion, the best thing would be to change this order.

jishnub commented:
Could you re-run the 2D and other dimensional benchmarks, but this time interpolating the function? This is what I obtain, for example, on a recent version:
```julia
julia> f2 = Fun(Chebyshev()^2, rand(100));

julia> @btime f2(0.2, 0.4);
  1.814 μs (23 allocations: 4.45 KiB)
```

benjione commented:
Thanks for catching that!

My changes should only apply to higher dimensional functions, d>2, but I did a mistake here. Currently the wrong iterator is called, which is using multiexponential from Combinatorics.jl and that is unfortunately slow. I will fix this, so that the original code is called for the 2d case and run the benchmark.

benjione commented:
I fixed the 2D case, it is now the same as before. Unfortunately, the higher dimensional cases are not very fast, because of the slow Stars and Bars algorithm I am using.... I already implemented an own version of the algorithm which is faster than the Julia implementation, maybe I will add this later on.

```
# all functions with 100 coefficients
917.500 ns (17 allocations: 4.34 KiB) #2D function evaluation
  382.679 μs (4675 allocations: 189.69 KiB) #3D function evaluation
  3.547 ms (16627 allocations: 1.02 MiB) #20D function evaluation
```

I will open a PR in the ApproxFunOrthogonalPolynomials for the multivariate case soon.

jishnub commented on src/Multivariate/ProductFun.jl:
```diff
@@ -5,6 +5,15 @@
+    space::SS
+    coefficients::Vector{T} 
+    iterator::TrivialTensorizer{d}
+    orders::Block
```
`Block` is an abstract type, this should ideally be concrete. Making this concrete will likely improve performance.

benjione replied:
fixed

benjione commented:
I have rewritten the iterator for better performances:
```
# all functions with 100 coefficients
1.007 μs (17 allocations: 4.34 KiB)  2D evaluation
  83.821 μs (1869 allocations: 98.86 KiB) 3D evaluation
  164.725 μs (6414 allocations: 201.72 KiB) 20D evaluation
```

jishnub commented on Project.toml:
```diff
@@ -47,8 +49,9 @@ StaticArrays = "0.12, 1.0"
 julia = "1.6"
 
 [extras]
+ApproxFunOrthogonalPolynomials = "b70543e2-c0d9-56b8-a290-0d4d6d4de211"
```
`ApproxFunOrthogonalPolynomials` i probably not needed anymore, as tests are moved to that repo?

benjione replied:
changed

jishnub commented on Project.toml:
```diff
@@ -51,4 +53,4 @@ Aqua = "4c88cf16-eb10-579e-8560-4a9242c79595"
 
 [targets]
-test = ["Aqua", "Random"]
+test = ["Aqua", "Random", "ApproxFunOrthogonalPolynomials"]
```
This should also be updated to remove `ApproxFunOrthogonalPolynomials`

jishnub commented on src/Multivariate/ProductFun.jl:
```diff
@@ -241,7 +243,6 @@ function coefficients(f::ProductFun,ox::Space,oy::Space)
 end
 
 (f::ProductFun)(x,y) = evaluate(f,x,y)
-(f::ProductFun)(x,y,z) = evaluate(f,x,y,z)
```
If I understand correctly, the new type `TensorIteratorFun` will handle nD functions efficiently. However, do we need to remove `ProductFun` methods?

benjione replied:
No. Actually, the `TensorIteratorFun` does not work for all bases, only for bases which have infinite coefficients, which are all except `ConstantSpace`, `SubSpace`, and some others. These spaces are referred to in the code as Trivial spaces, and hence `TrivialTensorizer` in TensorSpace.jl. The creation of a tensor for these kind of spaces is just simpler than for others and more efficient.
I will add a check in the constructor of `TensorIteratorFun` that the spaces are really only of this type.

`ProductFun` also handles "nonTrivialSpaces" but only works in 2D. I therefore deleted this method for 3D, the program currently just hangs when trying to evaluate this function. 

The reason `ProductFun` is faster for 2D polynomials than the `TensorIteratorFun`, is that for the 2D case there is a simpler method than Stars and Bars to convert the coefficient vector to a tensor. But `ProductFun` actually builds the full tensor, which is not efficient in the high dimensional setting. Therefore, in `TensorIteratorFun` I never actually build the full tensor but calculate the result directly from the `Tensorizer` iterator.

Also, currently the "detensorizer" for the nD case is missing. For this reason some things working in 2D will not work in 3D (I think multiplication and some others, but addition works).
But I think adding this is maybe better for another PR, since this already contains a lot of changes and it is not breaking with anything, since this also did not work before.

jishnub replied:
It'll be better if you comment the code out instead of deleting it, and leave this explanation as a comment

benjione commented:
@jishnub Are you going to accept this PR?
